core/drivers/crypto/crypto_api/drvcrypt_init.c

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   This driver interfaces TEE Internal API by implementing
+ *         all crypto_* functions.
+ */
+#include <crypto/crypto.h>
+#include <drvcrypt.h>
+#include <initcall.h>
+#include <trace.h>
+
```

Comment:
I'm not sure the interface implemented in this file is needed. It only seems safe to call during early boot, before exiting to normal world the first time. Type safety is lost. The caller of `drvcrypt_getmod()` must know the type of the point it gets. All this could be replaced by:
``` C
const struct drvcrypt_hash *drvcrypt_hash_ops;
const struct drvcrypt_mac *drvcrypt_mac_ops;
const struct drvcrypt_cipher *drvcrypt_cipher_ops;
...
```
with retained type safety. Perhaps I'm missing something.

If you think it's important to keep this, do so. Once all the PRs are merged we can evaluate this again.

Reply:
This interface is to register to HW Driver module in the crypt_algo array. The registration is done during the HW Driver initialization function of the HW capabilities. To be generic and to prevent optional compilation per device type, we read HW registers to know if CAAM support or not the algorithm to register. If not, we keep NULL in the crypt_algo and so the HW driver is not called.
For sure you can not see to need with the HASH but for other operation it will be usefull